src/authn/index.js

```diff
@@ -437,8 +438,27 @@ function setup(app) {
  * @type {expressMiddleware}
  */
 function authnWithToken(req, res, next) {
-  const authenticate = passport.authenticate(STRATEGY_BEARER, (err, user) => {
+  const authenticate = passport.authenticate(STRATEGY_BEARER, (err, user, challenge, status) => {
+    /* Passport strategies have four result states.  Handle all four, as the
+     * STRATEGY_BEARER can produce all of them.
+     */
+
+    // error: something went wrong internally; can't be remediated by client
     if (err) return next(err);
+
+    // failure: authn attempted but failed; may be remediated by client
+    if (!user && (challenge || status)) {
+      if (challenge) {
+        res.set("WWW-Authenticate", challenge);
+      }
```

Comment:
From `passport.authenticate`'s [docstring](https://github.com/jaredhanson/passport/blob/fc0fdc804fb5dbf99c510b6b43fa05dded9f4f48/lib/middleware/authenticate.js#L38C20-L38C20):

> An optional `info` argument will be passed, containing additional details provided by the strategy's verify callback - this could be information about a successful authentication or a challenge message for a failed authentication.

It looks like `challenge` is assuming a special case of `info`. Would the `if (challenge)` condition ever fail? If not, would it be better to verify that it's actually a challenge message (e.g. based on your example, `challenge.startsWith("error")`)?

Reply:
Nod. It is indeed assuming a special case of `info`, gated by the `!user` condition. It's not always truthy, however: if the request's `Authorization` header when split on a space contains fewer than or more than 2 parts.

https://github.com/nextstrain/nextstrain.org/blob/1557c827003de7bfcbee0e07f9f70687c7270b1f/src/authn/bearer.js#L101-L113

The `this.fail(400)` call there results in a callback invocation like:

    // err, user, challenge, status
    callback(null, false, undefined, 400);

Calls to `this.fail(…)` are collected and only invoke the callback if all strategies fail:

https://github.com/jaredhanson/passport/blob/cfdbd4a762b51e339ebfea931d65bccbbde53282/lib/middleware/authenticate.js#L107-L116

Reply:
Thanks for the explanation!